## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [69]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0    $1096.80    $762.71        $7.50  105            1  
1      $24.95     $15.09        $1.49   21            0  
2      $13.52      $7.86        $1.68   39            6  
3      $36.42     $24.85        $1.23   29            3  
4     $195.10    $108.17       $46.43   20            1

In [10]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [40]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [74]:
# Use this space to do any additional research


# and familiarize yourself with the data
#unit_price_df = ('unit_price')


In [75]:
# What three item categories had the most entries?
df.loc[df['category'] ==
"consumables"].value_counts("subcategory").head


top_categories = df['category'].value_counts().nlargest(3)
print("Top three item categories:")
print(top_categories) 

Top three item categories:
category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [52]:
# For the category with the most entries, which subcategory had the most entries?
most_common_category = top_categories.index[0]
subcategory_counts = df[df['category'] == most_common_category]['subcategory'].value_counts()
most_common_subcategory = subcategory_counts.idxmax()
print(f"Subcategory with the most entries in the '{most_common_category}' category: {most_common_subcategory}")

Subcategory with the most entries in the 'consumables' category: bathroom supplies


In [64]:
# Which five clients had the most entries in the data?
top_clients = df['client_id'].value_counts().nlargest(5)
print("Top five clients:")
print(top_clients)

Top five clients:
client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [205]:
# Store the client ids of those top 5 clients in a list.
top_client_ids = top_clients.index.tolist()
print("Top five client IDs:")
top_client_ids


Top five client IDs:


[78965, 85931, 38473, 86721, 84344]

In [23]:
# How many total units (the qty column) did the client with the most entries order order?
most_common_client_id = top_client_ids[0]
total_units_ordered = df[df['client_id'] == most_common_client_id]['qty'].sum()
print(f"Total units ordered by client {most_common_client_id}: {total_units_ordered}")



Total units ordered by client 33615: 64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [179]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
# Calculate the subtotal by multiplying unit_price and qty
df['subtotal'] = df['unit_price'] * df['qty']
pd.set_option('display.float_format', '${:.2f}'.format)
df.pivot_table(index='unit_price',
               columns='qty',
               values='subtotal',
               aggfunc='sum')
# Now your DataFrame 'df' has an additional 'subtotal' column
df.head() # Display the first few rows to verify


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art   $1096.80   
1          20        2023  ...  consumables                pens     $24.95   
2           4        2023  ...     software  project management     $13.52   
3          17        2023  ...  consumables                pens     $36.42   
4          21        2023  ...  consumables                misc    $195.10   

   unit_cost  unit_weight  qty  line_number  Shipping Price   subtotal  \
0    $762.71        $7.50  105            1              10 $115164.00   
1     $15.09        $1.49   21            0              10    $523.95   
2      $7.86        $1.68   39            6              10    $527.28   
3     $24.85        $1.23   29            3              10   $1056.18   
4    $108.17       $46.43   20            1              10   $3902.00   

   Total Price  
0   $125827.60  
1      $583.34  
2      $586.98  
3     $1164.80  
4     $4273.86  

[5 rows x 21 columns]

In [177]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
shipping_price = df['unit_weight'] 
# Calculate the shipping price based on the weight of the items
shipping_price = [df['Shipping Price'] == df['unit_weight'].apply(lambda w: 7 if w > 50 else 10)]

shipping_price = pd.DataFrame(shipping_price)

shipping_price.T




0
0      True
1      True
2      True
3      True
4      True
...     ...
54634  True
54635  True
54636  True
54637  True
54638  True

[54639 rows x 1 columns]

In [186]:
# Create a column for the total price using the subtotal and the shipping price 
#along with a sales tax of 9.25%


#df['Total Price'] = (df['subtotal'] + df['Shipping Price']) * 1.0925

df['unit_price'] = round((df['subtotal'] + df['Shipping Price']) * 1.0925, 2)
df[['subtotal', 'Shipping Price', 'unit_price']].head(3)

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art $125827.60   
1          20        2023  ...  consumables                pens    $583.34   
2           4        2023  ...     software  project management    $586.98   
3          17        2023  ...  consumables                pens   $1164.80   
4          21        2023  ...  consumables                misc   $4273.86   

   unit_cost  unit_weight  qty  line_number  Shipping Price   subtotal  \
0    $762.71        $7.50  105            1              10 $115164.00   
1     $15.09        $1.49   21            0              10    $523.95   
2      $7.86        $1.68   39            6              10    $527.28   
3     $24.85        $1.23   29            3              10   $1056.18   
4    $108.17       $46.43   20            1              10   $3902.00   

   Total Price  
0   $125827.60  
1      $583.34  
2      $586.98  
3     $1164.80  
4     $4273.86  

[5 rows x 21 columns]

In [188]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost'] = df['unit_cost'] * df['qty'] + df['Shipping Price']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art $125827.60   $762.71   
1          20        2023  ...                pens    $583.34    $15.09   
2           4        2023  ...  project management    $586.98     $7.86   
3          17        2023  ...                pens   $1164.80    $24.85   
4          21        2023  ...                misc   $4273.86   $108.17   

   unit_weight  qty  line_number  Shipping Price   subtotal  Total Price  \
0        $7.50  105            1              10 $115164.00   $125827.60   
1        $1.49   21            0              10    $523.95      $583.34   
2        $1.68   39            6              10    $527.28      $586.98   
3        $1.23   29            3              10   $1056.18     $1164.80   
4       $46.43   20            1              10   $3902.00     $4273.86   

   line_cost  
0  $80094.55  
1    $326.89  
2    $316.54  
3    $730.65  
4   $2173.40  

[5 rows x 22 columns]

In [168]:
# Create a column for the profit of each line using line cost and line price
df["subtotal"] =  df['unit_price'] * df['qty']

df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art   $1096.80   
1          20        2023  ...  consumables                pens     $24.95   
2           4        2023  ...     software  project management     $13.52   
3          17        2023  ...  consumables                pens     $36.42   
4          21        2023  ...  consumables                misc    $195.10   

   unit_cost  unit_weight  qty  line_number  Shipping Price   subtotal  \
0    $762.71        $7.50  105            1              10 $115164.00   
1     $15.09        $1.49   21            0              10    $523.95   
2      $7.86        $1.68   39            6              10    $527.28   
3     $24.85        $1.23   29            3              10   $1056.18   
4    $108.17       $46.43   20            1              10   $3902.00   

   Total Price  
0   $125827.60  
1      $583.34  
2      $586.98  
3     $1164.80  
4     $4273.86  

[5 rows x 21 columns]

In [210]:
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'Shipping Price', 'subtotal', 'Total Price',
       'line_cost'],
      dtype='object')

In [160]:
## Part 3: Confirm your work
#You have email receipts showing that the total prices for 3 orders. 
#Confirm that your calculations match the receipts. Remember, each order has multiple lines.

#Order ID 2742071 had a total price of \$152,811.89

#Order ID 2173913 had a total price of \$162,388.71

#Order ID 6128929 had a total price of \$923,441.25


In [216]:
# Check your work using the totals above
orders = [
    {"name": "Order ID 2742071", "price": 152811.89},
    {"name": "Order ID 2173913", "price": 162388.71},
    {"name": "Order ID 6128929", "price": 923441.25}
]
total_price = 0.0 
for order in orders:
    total_price += order["price"]
f"The total price for all orders is ${total_price:.2f}"


'The total price for all orders is $1238641.85'

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [218]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
top_clients = df.groupby(['client_id'])['Total Price'].sum().nlargest(5)
top_clients.head()

client_id
78965   $650090783.19
85931   $226358717.31
38473   $178757235.64
86721   $110980847.39
84344   $101463966.26
Name: Total Price, dtype: float64

In [17]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
summary_df.sort_values(by='Total Profit', ascending=False, inplace=True)

summary_df = pd.DataFrame({
    'Client ID': top_clients.index,
    'Total Units Purchased': top_clients.values,
    'Total Shipping Price': df.groupby('client_id')['Shipping Price'].sum(),
    'Total Revenue': df.groupby('client_id')['Total Price'].sum(),
    'Total Profit': df.groupby('client_id')['Profit'].sum())



client_id     qty  shipping_price   line_price    line_cost  line_profit
0      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61
1      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81
2      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
3      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
4      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31

In [18]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 


# Define a function that converts a dollar amount to millions.


# Apply the currency_format_millions function to only the money columns. 


# Rename the columns to reflect the change in the money format. 
summary_df.rename(columns={'Client ID': 'Client', 'Total Units Purchased': 'Units Purchased'}, inplace=True)



Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
0      33615   64313             1.828985                  8.377309   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
3      38378   73667             3.429455                 12.906551   
4      24741  239862             5.126448                 82.268892   

   Total Cost (millions)  Total Profit (millions)  
0               6.175314                 2.201995  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
3               9.634721                 3.271830  
4              45.688900                36.579992

In [19]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
def convert_to_millions(value):
    return f"${value / 1e6:.2f}M"
summary_df['Total Revenue'] = summary_df['Total Revenue'].apply(convert_to_millions)
summary_df['Total Profit'] = summary_df['Total Profit'].apply(convert_to_millions)

Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
4      24741  239862             5.126448                 82.268892   
3      38378   73667             3.429455                 12.906551   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
0      33615   64313             1.828985                  8.377309   

   Total Cost (millions)  Total Profit (millions)  
4              45.688900                36.579992  
3               9.634721                 3.271830  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
0               6.175314                 2.201995